lib/async_hooks.js

```diff
@@ -225,80 +223,65 @@ const storageHook = createHook({
 class AsyncLocalStorage {
   constructor() {
     this.kResourceStore = Symbol('kResourceStore');
-    this.enabled = false;
   }
 
   disable() {
-    if (this.enabled) {
-      this.enabled = false;
-      // If this.enabled, the instance must be in storageList
-      storageList.splice(storageList.indexOf(this), 1);
-      if (storageList.length === 0) {
-        storageHook.disable();
-      }
+    storageList.splice(storageList.indexOf(this), 1);
+    if (storageList.length === 0) {
+      storageHook.disable();
     }
   }
 
   // Propagate the context from a parent resource to a child one
   _propagate(resource, triggerResource) {
     const store = triggerResource[this.kResourceStore];
-    if (this.enabled) {
-      resource[this.kResourceStore] = store;
-    }
+    resource[this.kResourceStore] = store;
   }
 
-  _enter() {
-    if (!this.enabled) {
-      this.enabled = true;
+  _enter(store) {
+    if (storageList.indexOf(this) === -1) {
```

Comment:
This check has `O(N)` cost while the previous one was `O(1)`. And it's located on a hot path (`.run*()` methods).

Reply:
Not _that_ much of a hot-path as the enter would normally only happen once per-request, and we could always switch `storageList` to be a `Set`. Not too worried about it. :thinking:

Reply:
> we could always switch `storageList` to be a `Set`.

Turning `storageList` into a `Set` is a bad idea, as the async hook iterates over it (which is the hottest path).

Reply:
True. The performance difference barely even registered after this change though. :man_shrugging:

Reply:
Is there an action we still think needs to be done here? As shown in the previous benchmarks, the performance difference doesn't even register. :man_shrugging:

Reply:
The penalty may become visible when multiple `AsyncLocalStorage` instance are in use, as the array size would grow.

Also, considering https://github.com/nodejs/node/pull/31950/files#r385535111, it looks like `enabled` flag should be kept. WDYT?

Reply:
Restored.